sync_diff_inspector/diff.go

```diff
@@ -49,118 +51,332 @@ type Diff struct {
+	allTablesMap := make(map[string][]string)
+
+	for _, schemaTables := range cfg.Tables {
+		if _, ok := allTablesMap[schemaStr(cfg.TargetDBCfg.InstanceID, schemaTables.Schema)]; ok {
+			continue
 		}
-		df.report.SetTableStructCheckResult(table.Name, eq1)
 
-		eq2, err := df.EqualTableData(table)
+		allTables, err := dbutil.GetTables(df.ctx, cfg.TargetDBCfg.Conn, schemaTables.Schema)
 		if err != nil {
-			log.Errorf("equal table error %v", err)
-			return errors.Trace(err)
+			return nil, errors.Errorf("get tables from %s.%s error %v", cfg.TargetDBCfg.InstanceID, schemaTables.Schema, errors.Trace(err))
+		}
+		allTablesMap[schemaStr(cfg.TargetDBCfg.InstanceID, schemaTables.Schema)] = allTables
+	}
+
+	for _, table := range cfg.TableCfgs {
+		for _, sourceTable := range table.SourceTables {
+			if _, ok := allTablesMap[schemaStr(sourceTable.InstanceID, sourceTable.Schema)]; ok {
+				continue
+			}
+
+			db, ok := df.sourceDBs[sourceTable.InstanceID]
+			if !ok {
+				return nil, errors.Errorf("unknow instance id %s", sourceTable.InstanceID)
+			}
+
+			allTables, err := dbutil.GetTables(df.ctx, db.Conn, sourceTable.Schema)
+			if err != nil {
+				return nil, errors.Errorf("get tables from %s.%s error %v", db.InstanceID, sourceTable.Schema, errors.Trace(err))
+			}
+			allTablesMap[schemaStr(db.InstanceID, sourceTable.Schema)] = allTables
 		}
-		if !eq2 {
-			log.Errorf("table %s's data is not equal", table.Name)
+	}
+
+	return allTablesMap, nil
+}
+
+// GetMatchTable returns all the matched table.
+func (df *Diff) GetMatchTable(db DBConfig, schema, table string, allTables []string) ([]string, error) {
+	tableNames := make([]string, 0, 1)
+
+	if table[0] == '~' {
+		tableRegex := regexp.MustCompile(fmt.Sprintf("(?i)%s", table[1:]))
+		for _, tableName := range allTables {
+			if !tableRegex.MatchString(tableName) {
+				continue
+			}
+			tableNames = append(tableNames, tableName)
 		}
-		df.report.SetTableDataCheckResult(table.Name, eq2)
+	} else {
+		tableNames = append(tableNames, table)
+	}
 
-		if eq1 && eq2 {
+	return tableNames, nil
+}
+
+// Close closes file and database connection.
+func (df *Diff) Close() {
+	if df.fixSQLFile != nil {
+		df.fixSQLFile.Close()
+	}
+
+	for _, db := range df.sourceDBs {
+		if db.Conn != nil {
+			db.Conn.Close()
+		}
+	}
+
+	if df.targetDB.Conn != nil {
+		df.targetDB.Conn.Close()
+	}
+}
+
+// Equal tests whether two database have same data and schema.
+func (df *Diff) Equal() (err error) {
+	defer df.Close()
+
+	df.wg.Add(1)
+	go func() {
+		df.WriteSqls()
+		df.wg.Done()
+	}()
+
+	reportResult := func(structEqual, dataEqual bool) {
+		if structEqual && dataEqual {
 			df.report.PassNum++
 		} else {
 			df.report.FailedNum++
 		}
 	}
 
+	for _, schema := range df.tables {
+		for _, table := range schema {
+			structEqual, err := df.CheckTableStruct(table)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			df.report.SetTableStructCheckResult(table.Table, structEqual)
```

Comment:
is possible there are multi tables in different schema have same `table.Table` , and later one overwrite earlier one?

Reply:
my mistake, done